### PR TITLE
Add final type hint to tests.unittest.

### PR DESCRIPTION
**Pull request by clokep**
This had a lot of fallout, I don't think it found any "bugs", but it did find a bunch of tests where we weren't testing what we thought we were.

Sorry this got pretty big, but I couldn't figure out a good way to split it up?

It should be reviewable commit-by-commit, which are done for directories.

### CODE REVIEW

File: tests/handlers/test_user_directory.py [hunk 1]
```diff
@@ -797,43 +797,6 @@ async def block_all(user_profile: ProfileInfo) -> bool:
         s = self.get_success(self.handler.search_users(u1, "user2", 10))
         self.assertEqual(len(s["results"]), 0)
 
-    def test_legacy_spam_checker(self) -> None:
```
Comment (clokep) [inline tests/handlers/test_user_directory.py:800]: I couldn't even figure out what this is attempting to test? It seems to be trying to test something from a really old way that modules were configured, but setting `spam_checker.spam_checker` is invalid?

File: tests/rest/client/test_register.py [hunk 2]
```diff
@@ -1166,12 +1166,15 @@ def test_background_job(self) -> None:
         """
         user_id = self.register_user("kermit_delta", "user")
 
-        self.hs.config.account_validity.startup_job_max_delta = self.max_delta
+        self.hs.config.account_validity.account_validity_startup_job_max_delta = (
+            self.max_delta
+        )
```
Comment (DMRobertson) [inline tests/rest/client/test_register.py:1171]: Was this just broken before?

**Review by DMRobertson — APPROVED** [review]
LGTM, though see the comment on the legacy spam checker stuff.

File: tests/handlers/test_e2e_keys.py [hunk 3]
```diff
@@ -451,8 +452,10 @@ def test_self_signing_key_doesnt_show_up_as_device(self) -> None:
         }
         self.get_success(self.handler.upload_signing_keys_for_user(local_user, keys1))
 
+        device_handler = self.hs.get_device_handler()
+        assert isinstance(device_handler, DeviceHandler)
```
[the remark this answers is not in the document]
Comment (clokep) [reply tests/handlers/test_e2e_keys.py:456]: Pretty much, yes. 👍

File: tests/rest/client/test_register.py (hunk 2 above)
Comment (clokep) [reply tests/rest/client/test_register.py:1171]: I think so.

File: tests/handlers/test_user_directory.py (hunk 1 above)
Comment (clokep) [reply tests/handlers/test_user_directory.py:800]: We can keep it, but we would need to adapt it and I couldn't figure out what a similar test is. I'll double check.

**Pull request merged**